annotated_line_chart/graphic.js

```diff
@@ -0,0 +1,418 @@
+    .append("svg")
+    .attr("width", chartWidth + margins.left + margins.right)
+    .attr("height", chartHeight + margins.top + margins.bottom)
+    .append("g")
+    .attr("transform", "translate(" + margins.left + "," + margins.top + ")");
+
+  /*
+   * Create D3 axes.
+   */
+  var xAxis = d3
+    .axisBottom()
+    .scale(xScale)
+    .ticks(ticksX)
+    .tickFormat(function(d, i) {
+      if (isMobile) {
+        return "\u2019" + fmtYearAbbrev(d);
+      } else {
+        return fmtYearFull(d);
+      }
+    });
+
+  var yAxis = d3
+    .axisLeft()
+    .scale(yScale)
+    .ticks(ticksY);
+
+  /*
+   * Render axes to chart.
+   */
+  chartElement
+    .append("g")
+    .attr("class", "x axis")
+    .attr("transform", makeTranslate(0, chartHeight))
+    .call(xAxis);
+
+  chartElement
+    .append("g")
+    .attr("class", "y axis")
+    .call(yAxis);
+
+  /*
+   * Render grid to chart.
+   */
+  var xAxisGrid = function() {
+    return xAxis;
+  };
+
+  var yAxisGrid = function() {
+    return yAxis;
+  };
+
+  chartElement
+    .append("g")
+    .attr("class", "x grid")
+    .attr("transform", makeTranslate(0, chartHeight))
+    .call(
+      xAxisGrid()
+        .tickSize(-chartHeight, 0, 0)
+        .tickFormat("")
+    );
+
+  chartElement
+    .append("g")
+    .attr("class", "y grid")
+    .call(
+      yAxisGrid()
+        .tickSize(-chartWidth, 0, 0)
+        .tickFormat("")
+    );
+
+  /*
+   * Render 0 value line.
+   */
+  if (min < 0) {
+    chartElement
+      .append("line")
+      .attr("class", "zero-line")
+      .attr("x1", 0)
+      .attr("x2", chartWidth)
+      .attr("y1", yScale(0))
+      .attr("y2", yScale(0));
+  }
+
+  /*
+   * Render lines to chart.
+   */
+  var line = d3
+    .line()
+    .x(function(d) {
+      return xScale(d[dateColumn]);
+    })
+    .y(function(d) {
+      return yScale(d[valueColumn]);
+    });
+
+  chartElement
+    .append("g")
+    .attr("class", "lines")
+    .selectAll("path")
+    .data(config.data)
+    .enter()
+    .append("path")
+    .attr("class", function(d, i) {
+      return "line " + classify(d.name);
+    })
+    .attr("stroke", function(d) {
+      return colorScale(d.name);
+    })
+    .attr("d", function(d) {
+      return line(d.values);
+    });
+
+  chartElement
+    .append("g")
+    .attr("class", "value")
+    .selectAll("text")
+    .data(config.data)
+    .enter()
+    .append("text")
+    .attr("x", function(d, i) {
+      var last = d.values[d.values.length - 1];
+      return xScale(last[dateColumn]) + 5;
+    })
+    .attr("y", function(d) {
+      var last = d.values[d.values.length - 1];
+      return yScale(last[valueColumn]) + 3;
+    });
+
+  /*
+   * Render annotations.
+   */
+  var annotation = chartElement
+    .append("g")
+    .attr("class", "annotations")
+    .selectAll("circle")
+    .data(config.annotations)
+    .enter();
+
+  annotation
+    .append("circle")
+    .attr("class", "dots")
+    .attr("cx", function(d) {
```

Comment:
For these kinds of functions, it's nice to express them with the fat arrow syntax to reduce visual noise: `d => xScale(d[valueColumn])`

Reply:
I realize now, of course, that I didn't necessarily do that in other places in the default line chart, but it's never to late to do as I say and not as I do.